Add Telegram Bot

mihvas opened the pull request:
Add a telegram bot for booking free cabins. The bot allows you to view vacant houses for a certain period, book it, view your applications and change them.

mihvas committed:
Add Telegram Bot

Add a telegram bot for booking free cabins. The bot allows you to view vacant houses for a certain period, book it, view your applications and change them.

mihvas commented:
Так как телеграм бот основывается на 3 задание, то часть изменений в этой ветке относится к 3 заданию. Я потом их вынесу и доработаю в отдельный pull-request для 3 задания.

praisethedeviI commented on docker-compose.yml:
```diff
@@ -23,36 +30,32 @@ services:
+      POSTGRES_PASSWORD: ${DB_PASSWORD}
     ports:
-      - "${DB_PORT}:5432"
+      - "${DB_PORT_IN}:${DB_PORT}"
```
Не понял немного мува. В .env.local как прописано подключение к постгресу: хост и порт?

praisethedeviI replied:
Хочу ответа. Какие DB_HOST и DB_PORT прописываешь в .env.local? Какой DATABASE_URL получается?

mihvas replied:
DB_HOST=db
DB_PORT=5432
DB_PORT_IN=5433

DATABASE_URL="pgsql://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}/${DB_NAME}?serverVersion=15&charset=utf8"

praisethedeviI commented on docker-compose.yml:
```diff
@@ -23,36 +30,32 @@ services:
+    image: postgres:latest
+    container_name: db_test_container
+    env_file:
+      - .env.test.local
```
Этого тоже не понял. Разве симфони сам не создает новую базу данных при старте юниттеста?

praisethedeviI commented on migrations/Version20250530051405.php:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20250530051405 extends AbstractMigration
```
Эта миграция лишняя

praisethedeviI commented on migrations/Version20250530051405.php:
```diff
@@ -0,0 +1,33 @@
+    {
+        // this down() migration is auto-generated, please modify it to your needs
+        $this->addSql(<<<'SQL'
+            CREATE SCHEMA public
```
https://github.com/doctrine/dbal/issues/1110

Если каждый раз пытается создаться схема в миграциях на down методе, то предлагаю добавить код с этого issue

Либо руками убирать этот sql

praisethedeviI commented on migrations/Version20250530053309.php:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20250530053309 extends AbstractMigration
```
тоже лишняя миграция

praisethedeviI commented on migrations/Version20250530053650.php:
```diff
@@ -0,0 +1,44 @@
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->addSql(<<<'SQL'
+            ALTER TABLE house ALTER date_start DROP NOT NULL
```
У тебя две одинаковые миграции: эта повторяет миграцию Version20250530051909. Надо оставить какую-то одну

А лучше вообще все миграции объединить в рамках одной задачи.  То есть в миграции Version20250530050756 просто не добавлять эти поля в таблицу house

praisethedeviI commented on phpunit.dist.xml:
```diff
@@ -15,6 +15,8 @@
         <ini name="error_reporting" value="-1" />
         <server name="APP_ENV" value="test" force="true" />
         <server name="SHELL_VERBOSITY" value="-1" />
+        <server name="DATABASE_URL" value="pgsql://symfony:symfony@db:5432/symfony?serverVersion=15&amp;charset=utf8"/>
```
Это надо в .env файлов тянуть

praisethedeviI commented on phpunit.dist.xml:
```diff
@@ -15,6 +15,8 @@
         <server name="APP_ENV" value="test" force="true" />
         <server name="SHELL_VERBOSITY" value="-1" />
+        <server name="DATABASE_URL" value="pgsql://symfony:symfony@db:5432/symfony?serverVersion=15&amp;charset=utf8"/>
+        <env name="KERNEL_CLASS" value="App\Kernel"/>
```
и это из .env файлов должно тянуться

praisethedeviI commented on src/Controller/BookingController.php:
```diff
@@ -1,12 +1,110 @@
 <?php
+//
```
Зачем нужен этот закомментированный код?

praisethedeviI commented on src/Controller/BookingController.php:
```diff
@@ -1,12 +1,110 @@
 use Symfony\Component\HttpFoundation\JsonResponse;
 use Symfony\Component\HttpFoundation\Request;
-use Symfony\Component\Routing\Attribute\Route;
+use Symfony\Component\Routing\Annotation\Route;
```
Почему поменялось?

praisethedeviI commented on src/Controller/BookingController.php:
```diff
@@ -25,22 +123,24 @@ public function createBooking(Request $request): JsonResponse
-        $comment = $data['comment'] ?? '';
+        $comment = $data['comment'] ?? null;
+        $userId = $data['userId'] ?? null;
+        $chadId = $data['chadId'] ?? null;
```
chatId ?:)

praisethedeviI commented on src/Controller/BookingController.php:
```diff
@@ -25,22 +123,24 @@ public function createBooking(Request $request): JsonResponse
-        $comment = $data['comment'] ?? '';
+        $comment = $data['comment'] ?? null;
+        $userId = $data['userId'] ?? null;
+        $chadId = $data['chadId'] ?? null;
```
Кстати, не понимаю почему эти параметры тут добавились

mihvas replied:
Для того чтоб моя сущности соответствовали 7-му заданию, я изменил их логику. Все остальное я также решил переделать под новую логику.

praisethedeviI replied:
Ну вообщем вопрос был больше такой: реально ли в контроллере надо ловить такие параметры, если они предназначены только через работу по телеграму? Я считаю, что нет, тк логика у контроллера все-таки друга, и там становится не понятно, что за чат.

mihvas replied:
Ну да это имеет смысл. Исправил.

praisethedeviI commented on src/Controller/BookingController.php:
```diff
@@ -51,45 +151,37 @@ public function updateBooking(Request $request): JsonResponse
-            http_response_code(404);
-            return $this->json(['error2' => 'Бронирование не найдено']);
+        $booking = $this->bookingService->findBookingById($id);
+        if (!$booking) {
```
`$booking !== null`. Или тут ожидается, что что-то еще может быть кроме null, но false?

praisethedeviI commented on src/Controller/BookingController.php:
```diff
@@ -51,45 +151,37 @@ public function updateBooking(Request $request): JsonResponse
-        return $this->json(['success' => true, 'message' => 'Комментарий обновлен']);
+        $bookingId = $this->bookingService->updateBooking($booking);
+        return $this->json(['success' => true,
+            'bookingId' => $bookingId, 'message' => 'Комментарий обновлен']);
```
Давай, когда такой массив пишем, писать каждый элемент с новой строки. Вот так:

```php
return $this->json([
     'success' => true,
     'bookingId' => $bookingId, 
     'message' => 'Комментарий обновлен',
]);
```

Есть вот такая [php-cs-fixer](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer) мегаполезная штука на такие случаи, если что

praisethedeviI commented on src/Controller/HouseController.php:
```diff
@@ -40,37 +41,50 @@ public function createHouse(Request $request): JsonResponse
+        return $this->json(['success' => true, 'houseId' => $id]);
     }
 
+//    #[Route('/update', name: 'update', methods: ['PUT'])]
```
давай без лишних комментов кода. Если этот просто копипаста старого кода, то оно все равно в гите останется

praisethedeviI commented on src/Controller/BookingController.php:
```diff
@@ -51,45 +151,37 @@ public function updateBooking(Request $request): JsonResponse
         }
 
-        $isDelete = $this->bookingService->deleteBooking((int)$id);
+        $deleted = $this->bookingService->deleteBooking((int)$id);
```
тут $deleted, а ниже в другом контроллере $isDeleted. Давай как-то консистентно везде такое писать. Предлагаю использовать $isDeleted

praisethedeviI commented on src/Controller/HouseController.php:
```diff
@@ -40,37 +41,50 @@ public function createHouse(Request $request): JsonResponse
         $address = $data['address'] ?? null;
         $price = $data['price'] ?? null;
-
+        $free = $data['free'] ?? true;
```
$isFree

praisethedeviI commented on src/Controller/TelegramController.php:
```diff
@@ -0,0 +1,89 @@
+            $state = $this->cache->getItem($key);
+            if ($state->isHit()) {
+                $data = $state->get();
+                $text1 = $data['action'];
```
давай осмысленно назовем переменную. text1 -- что-то непонятное. Почему 1?

praisethedeviI commented on src/Service/TelegramService.php:
```diff
@@ -0,0 +1,462 @@
+        $this->cache = $cache;
+    }
+
+    public function safeText(string $text): string
```
как будто этот метод должен быть приватным. Его же только методы класса используют?

praisethedeviI commented on src/Service/TelegramService.php:
```diff
@@ -0,0 +1,462 @@
+        $this->telegram = new Api($token);
+        $this->em = $em;
+        $this->cache = $cache;
+    }
```
Если что с php8.1 можно в конструкторе объявлять свойства. В твоем случае будет:

```php
private readonly ParameterBagInterface $params, 
private readonly EntityManagerInterface $em, 
private readonly CacheInterface $cache,
```

readonly, потому что инстансы на этих переменных не надо будет менять после их инициализации

praisethedeviI commented on src/Service/TelegramService.php:
```diff
@@ -0,0 +1,462 @@
+                $this->messageBookingList($chatId, "Свободных домиков на эти даты нет.");
+                return;
+            }
+            $index1 = $index;
```
снова 1 в названии переменной. Переименовать надо

praisethedeviI commented on src/Service/TelegramService.php:
```diff
@@ -0,0 +1,462 @@
+            }
+            $index1 = $index;
+            if (!is_null($house)) {
+                foreach ($houses as $i => $h) {
```
и тут еще порезанные названия переменных

praisethedeviI commented on src/Service/TelegramService.php:
```diff
@@ -0,0 +1,462 @@
+use Telegram\Bot\Api;
+use \Cake\Chronos\Chronos;
+
+class TelegramService
```
Могу узнать весь флоу работы с телеграмом? Что, когда, почему и откуда

И плюс более точные вопросы:
* Не понимаю, зачем кэширование. И если кэширование, то где?
* Почему некоторые поля у сущности booking стали опциональными?

mihvas replied:
![image](https://github.com/user-attachments/assets/41c02ab5-f2b2-46eb-877d-e0a4067f1e55)
 Пользователь может выбрать дату согласно формату и ему присылается список свободных домиков.
![image](https://github.com/user-attachments/assets/21c6ef4c-c74d-4050-8989-7d96dcb8b430)
При нажатии на домик он его бронирует. Пользователь может посмотреть список всех своих заявок.
![image](https://github.com/user-attachments/assets/a95eedee-c130-4350-9299-c731bd0b4044)
При нажатии на заявку пользователь может произвести разные действия с ней. Тут как раз и используется кэширование. Оно нужно для того, чтоб понимать что пользователь вводит: телефон, комментарий или даты для поиска домиков. Если пользователь передумал изменять что-то в заявки, то он может отменить это действие.

В 7 задании в сущность booking добавились поля chatId и userId, поэтому поле phone стало необязательным, а скорее дополнительным для связи пользователя и арендодателя.

praisethedeviI commented on tests/Repository/BookingRepositoryTest.php:
```diff
@@ -1,84 +1,123 @@
+        $this->userId = 111;
+        $this->booking1 = new Booking($this->userId, 1112,'Comment 1', '71234567890');
+        $this->booking2 = new Booking(222, 2222,'Comment 2', '77777777777');
+//
```
убрать

praisethedeviI commented on src/Service/TelegramService.php:
```diff
@@ -0,0 +1,462 @@
+
+    public function handleBookingRequest(int $chatId, int $userId, string $username, string $callbackData): void
+    {
+        if (str_starts_with($callbackData, 'book_')) {
```
Честно говоря вот эти if elseif  -- не оч. На самом деле это ведь своего рода маршрутизация. Тут надо только обработать эти строки (или лучше вообще в контроллере резольв путей телеграма делать), при каждом кейсе вызывать какой-то метод из этого сервиса

Тогда вот этот метод не будет таким нагроможденным

mihvas committed:
Refactor Telegram Bot logic

Transfer the callback logic from the service to the controller, change the names of variables and delete unnecessary comments.

praisethedeviI commented on migrations/Version20250606090058.php:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20250606090058 extends AbstractMigration
```
все еще есть лишняя миграция

praisethedeviI commented on migrations/Version20250606091638.php:
```diff
@@ -0,0 +1,53 @@
+    public function down(Schema $schema): void
+    {
+        // this down() migration is auto-generated, please modify it to your needs
+        $this->addSql(<<<'SQL'
+            CREATE SCHEMA public
+        SQL);
```
не убрал этот sql

praisethedeviI commented on src/Service/TelegramService.php:
```diff
@@ -0,0 +1,456 @@
+                                private readonly CacheInterface         $cache)
+    {
+        $token = $params->get('telegram_bot_token');
+        $this->telegram = new Api($token);
```
У тебя это свойство объявляется динамически теперь. Добавь свойство private readonly Api $telegram в начале класса

mihvas committed:
Fix Booking logic

Delete unnecessary Booking entity parameters from bookingController and bookingService. Fix minor bugs.
